apple/swift-docc-render
# Start documenting components with DocC using in-source comments

Bug/issue #, if applicable: 117215034

## Summary

This adds a script to generate a "symbol graph" JSON file, which can be used to document Swift-DocC-Render using Swift-DocC itself.

Specifically, this script utilizes [tree-sitter][tree-sitter] parsers to inspect each `.vue` file in the `src/components` directory and emit a "symbol" node in the graph corresponding to each Vue component. By default, a title and declaration are created for every component. If a component defines its props, a "Parameters" section will be automatically generated to list the type and optionality of each prop. Additionally, you can add abstract/overview/parameter documentation by using a multi-line comment immediately preceding the export statement for the component, using the normal DocC [structure][docc-structure] and [formatting][docc-formatting].

With the following command, the symbol graph is generated as part of invoking Swift-DocC to build the documentation, and it can be previewed at http://localhost:8000/documentation/swiftdoccrender.

```shell
npm run docs
```

Now, this will include pages for every component in addition to the articles in `SwiftDocCRender.docc`. The existing catalog can also be used to document components using [extension files][docc-extension-files] like normal Swift symbols. As an example, the `Reference` component is documented using a mix of an extension file and its in-source documentation comment.

**Example screenshot:**
<img width="1436" alt="Example screenshot of component documentation page" src="https://github.com/apple/swift-docc-render/assets/212918/7e2d06c5-dba4-4a59-9f42-cb8d9802f5c9">

This example page is generated from the in-source documentation for the [`WordBreak`][component-word-break] component.

### Future directions and other notes

I think the ability to document Swift-DocC-Render using Swift-DocC is something that should motivate us to write better documentation for components going forward.

I have done very minimal documentation/curation of existing components in this PR, but I think this is something that will naturally grow and evolve over time as we work on the app.

I think it makes sense to utilize a GitHub action to eventually publish the output of `npm run docs:build` as a GitHub Pages website for this repo and have it automatically update as documentation changes.

The way tree-sitter is used here is something that could potentially be done in a more generic way that allows for documenting code in other languages and not just Vue projects in a future world, but there will likely need to be changes made to the symbolgraph format to make it work in a more language-agnostic way.

[tree-sitter]: https://tree-sitter.github.io/tree-sitter/
[docc-structure]: https://www.swift.org/documentation/docc/writing-symbol-documentation-in-your-source-files
[docc-formatting]: https://www.swift.org/documentation/docc/formatting-your-documentation-content
[docc-extension-files]: https://www.swift.org/documentation/docc/adding-supplemental-content-to-a-documentation-catalog#Add-Extension-Files-to-Append-to-or-Override-Source-Documentation-Comments
[component-word-break]: https://github.com/mportiz08/swift-docc-render/blob/9680b2170951e682807ec3ab785d52da62544369/src/components/WordBreak.vue#L12

## Dependencies

To run the `npm run docs` command, you will need to have a version of Swift-DocC installed and available on your system, either from Xcode or a Swift toolchain.

## Testing

Steps:
1. Run `npm run docs` and preview the docs
2. Verify that you can make changes to component comments and/or markdown files and run `npm run docs` again to see those changes reflected in the resulting web pages.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.
